### PR TITLE
Implement Inlinebots

### PR DESCRIPTION
**Pull request by jh0ker**
Implement Telegram Bot API changes from January 4, 2016 and January 14, 2016
Related to #146

### CODE REVIEW

File: telegram/bot.py [hunk 1]
```diff
@@ -586,6 +586,56 @@ def sendChatAction(self,
...
+        data = {'inline_query_id': inline_query_id,
+                'results': results}
+
+        if cache_time is not None:
+            data['cache_time'] = cache_time
+        if is_personal is not None:
+            data['is_personal'] = is_personal
+        if next_offset is not None:
+            data['next_offset'] = next_offset
+
```
Comment (tsnoam) [inline telegram/bot.py]: who may call this method? what will happen if `cache_time` is not int? if `is_personal` is not bool and `next_offset` is not str?
can it cause an exception raised at a later time? (will cause confusion on the real origin of the problem)

File: telegram/choseninlineresult.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        from_user (:class:`telegram.User`):
+        query (str):
+
+    """
+
+    def __init__(self,
+                 result_id,
+                 from_user,
+                 query,
+                 **kwargs):
```
Comment (tsnoam) [inline telegram/choseninlineresult.py]: The line feed after each argument... Is it from the google style guide?

File: telegram/dispatcher.py [hunk 3]
```diff
@@ -254,21 +257,23 @@ def processUpdate(self, update, context=None):
...
 
+            # Telegram update (message)
+            else:
+                self.dispatchTelegramMessage(update, context)
+                handled = True
+        elif isinstance(update, Update) and \
+                (update.inline_query is not None or
+                 update.chosen_inline_result is not None):
+                self.dispatchTelegramInline(update, context)
+                handled = True
```
Comment (tsnoam) [inline telegram/dispatcher.py]: I think that this method (`processUpdate()`)  became too complex and hard to follow the flow (fwiw, it wasn't simple before neither :-\ )
I suggest refactoring it and possibly splitting it into sub-methods to handle the internal flows.

File: telegram/inlinequeryresult.py [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser Public License for more details.
+#
+# You should have received a copy of the GNU Lesser Public License
+# along with this program.  If not, see [http://www.gnu.org/licenses/].
+
+"""
+This module contains the classes that represent Telegram InlineQueryResults
```
Comment (tsnoam) [inline telegram/inlinequeryresult.py]: can you add a link to the relevant section in the api documentation?

File: telegram/inlinequeryresult.py [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+        thumb_url (Optional[str]):
+        thumb_width (Optional[int]):
+        thumb_height (Optional[int]):
+    """
+
+    def __init__(self,
+                 id,
+                 title,
+                 message_text,
+                 **kwargs):
```
Comment (tsnoam) [inline telegram/inlinequeryresult.py]: why **kwargs and not arguments with default value as None?

File: telegram/inlinequeryresult.py [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+        # Required
+        super(InlineQueryResultArticle, self).__init__('article', id)
+        self.title = title
+        self.message_text = message_text
+
+        # Optional
+        self.parse_mode = kwargs.get('parse_mode', '')
+        self.disable_web_page_preview = kwargs.get('disable_web_page_preview',
+                                                   False)
+        self.url = kwargs.get('url', '')
```
Comment (tsnoam) [inline telegram/inlinequeryresult.py]: What will happen if `url` is not a string? (similar question to the other kwargs)

File: telegram/inlinequeryresult.py [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+        id (str): Unique identifier for this result, 1-64 Bytes
+
+    """
+
+    def __init__(self,
+                 type,
+                 id):
+        # Required
+        self.type = str(type)
+        self.id = str(id)
```
Comment (tsnoam) [inline telegram/inlinequeryresult.py]: nitpick: `type` & `id` are python bulitins. you might want to consider this instead:
```
def __init__(self,
             type_,
             id_):
    self.type = type_
    self.id = id_
```

File: telegram/inlinequeryresult.py [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+
+        Returns:
+            telegram.InlineQueryResultArticle:
+        """
+        if not data:
+            return None
+
+        return InlineQueryResultArticle(**data)
+
+
```
Comment (tsnoam) [inline telegram/inlinequeryresult.py]: all above comments applies for the following classes

File: tests/test_choseninlineresult.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see [http://www.gnu.org/licenses/].
+
+"""This module contains a object that represents Tests for Telegram
+ChosenInlineResult"""
+
+import os
```
Comment (tsnoam) [inline tests/test_choseninlineresult.py]: `os` is not used

File: tests/test_choseninlineresult.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+import unittest
+import sys
+sys.path.append('.')
```
Comment (tsnoam) [inline tests/test_choseninlineresult.py]: are you sure that this is necessary? afaik, current working directory is in `sys.path` by definition. am i missing something?

File: tests/test_choseninlineresult.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+        self.json_dict = {
+            'result_id': self.result_id,
+            'from': self.from_user.to_dict(),
+            'query': self.query
+        }
+
+    def test_choseninlineresult_de_json(self):
+        """Test ChosenInlineResult.de_json() method"""
+        print('Testing ChosenInlineResult.de_json()')
```
Comment (tsnoam) [inline tests/test_choseninlineresult.py]: you don't really need the docstring and print here.
when running `nosetests -v` it will print the method name by default and imho this is the most informative for the developer.

File: tests/test_choseninlineresult.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+        result = telegram.ChosenInlineResult.de_json(self.json_dict)
+
+        self.assertEqual(result.result_id, self.result_id)
+        self.assertEqual(result.from_user.to_dict(), self.from_user.to_dict())
```
Comment (tsnoam) [inline tests/test_choseninlineresult.py]: use `self.assertDictEqual` instead

File: tests/test_inlinequery.py [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+        }
+
+    def test_inlinequery_de_json(self):
+        """Test InlineQuery.de_json() method"""
+        print('Testing InlineQuery.de_json()')
+
+        inlinequery = telegram.InlineQuery.de_json(self.json_dict)
+
+        self.assertEqual(inlinequery.id, self.id)
+        self.assertEqual(inlinequery.from_user.to_dict(), self.from_user.to_dict())
```
Comment (tsnoam) [inline tests/test_inlinequery.py]: 1. use `self.assertDictEqual` instead
1. nitpick: line is longer than 79 chars

File: tests/test_inlineresult.py [hunk 14]
```diff
@@ -0,0 +1,442 @@
...
+        self.disable_web_page_preview = True
+        self.url = 'url'
+        self.hide_url = True
+        self.description = 'description'
+        self.thumb_url = 'thumb url'
+        self.thumb_height = 10
+        self.thumb_width = 15
+
+        self.json_dict = {
+            'type': 'article',
```
Comment (tsnoam) [inline tests/test_inlineresult.py]: consider using a variable for `'article'` (i.e.:
```
self.type = 'article'
...
self.json_dict = {
    'type': self.type,
...
```
then later on you can call `assertEqual` with `self.type`

File: tests/test_inlineresult.py [hunk 15]
```diff
@@ -0,0 +1,442 @@
...
+        self.assertEqual(article.message_text, self.message_text)
+        self.assertEqual(article.parse_mode, self.parse_mode)
+        self.assertEqual(article.disable_web_page_preview,
+                         self.disable_web_page_preview)
+        self.assertEqual(article.url, self.url)
+        self.assertEqual(article.hide_url, self.hide_url)
+        self.assertEqual(article.description, self.description)
+        self.assertEqual(article.thumb_url, self.thumb_url)
+        self.assertEqual(article.thumb_height, self.thumb_height)
+        self.assertEqual(article.thumb_width, self.thumb_width)
```
Comment (tsnoam) [inline tests/test_inlineresult.py]: can `self.assertDictEqual` be used here instead of all the `assertEqual`

**Comment by tsnoam** [comment]
i see that the coverage had decreased. i couldn't see in travis why (it claims that the files don't exist). on the next CI check lets make sure that the important parts are covered properly.

File: telegram/dispatcher.py (hunk 3 above)
Comment (jh0ker) [inline telegram/dispatcher.py]: I agree with you on that ^^

File: tests/test_choseninlineresult.py (hunk 10 above)
Comment (jh0ker) [inline tests/test_choseninlineresult.py]: I am actually not sure about that either, but this has been done in all other tests as well, so I just assumed it would be necessary.

File: tests/test_choseninlineresult.py (hunk 11 above)
Comment (jh0ker) [inline tests/test_choseninlineresult.py]: As above, this is done in other test packages, so I went with it.

File: tests/test_choseninlineresult.py (hunk 12 above)
Comment (jh0ker) [inline tests/test_choseninlineresult.py]: I did not know about this. It's done in other tests like that as well, maybe we can change it in other places as well.

File: tests/test_inlinequery.py (hunk 13 above)
Comment (jh0ker) [inline tests/test_inlinequery.py]: Weird that travis does not notice it

File: tests/test_inlineresult.py (hunk 14 above)
Comment (jh0ker) [inline tests/test_inlineresult.py]: agreed

**Comment by jh0ker** [comment]
I think the coverage decreases because there is no test for `answerInlineQuery`, since I was unsure about how to test it automatically

File: telegram/inlinequeryresult.py (hunk 7 above)
Comment (jh0ker) [inline telegram/inlinequeryresult.py]: I noticed, but I did it like this cause it is done in other places of the code (eg User.py), for consistency.

File: telegram/inlinequeryresult.py (hunk 5 above)
Comment (jh0ker) [inline telegram/inlinequeryresult.py]: I could say that it would make the code unnecessarily big, but I guess I was also kind of lazy about it. As with a lot of other comments, this is done throughout the library, so I went with it.

File: telegram/bot.py (hunk 1 above)
Comment (jh0ker) [inline telegram/bot.py]: It will either cause the JSON library to raise an error (`<type> is not JSON-serializable`) or an API error, if it goes through to telegram servers.

File: telegram/inlinequeryresult.py (hunk 6 above)
Comment (jh0ker) [inline telegram/inlinequeryresult.py]: (Similar answer)

File: telegram/choseninlineresult.py (hunk 2 above)
Comment (jh0ker) [inline telegram/choseninlineresult.py]: Seemed to be library convention as it is done in most other places, too.

File: tests/test_inlineresult.py (hunk 15 above)
Comment (jh0ker) [inline tests/test_inlineresult.py]: Yes

**Comment by jh0ker** [comment]
Travis is failing due to network errors

File: telegram/bot.py (hunk 1 above)
Comment (tsnoam) [inline telegram/bot.py]: In that case I think that enforcing the proper types is required. Helps identifying errors in the proper place and avoids user questions later.
For `int` and `bool` it can be done just as simple as `int(cache_time)` & `bool(is_personal)`.
About `str` there's the unicode compatibility in python 2.x. So checking `isinstance(next_offset, basestring)` and raising `ValueError` is in order.

File: telegram/inlinequeryresult.py (hunk 5 above)
Comment (tsnoam) [inline telegram/inlinequeryresult.py]: imho, using *args & **kwargs must be used with care. i find the most useful when you wrap another function (and even then it can be misleading and a pitfall for typos). when it's for "your own" function you should be explicit and specify each argument in the function prototype.

File: telegram/choseninlineresult.py (hunk 2 above)
Comment (tsnoam) [inline telegram/choseninlineresult.py]: oh well.

File: telegram/inlinequeryresult.py (hunk 6 above)
Comment (tsnoam) [inline telegram/inlinequeryresult.py]: (similar answer as well :) )

File: telegram/inlinequeryresult.py (hunk 7 above)
Comment (tsnoam) [inline telegram/inlinequeryresult.py]: ok

File: tests/test_choseninlineresult.py (hunk 11 above)
Comment (tsnoam) [inline tests/test_choseninlineresult.py]: you don't have to keep going with that then :-\
after all, these are unitests and they should be as most convenient for the developer.
